lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1672 @@
+			BtreeEntry::Occupied(mut entry) => {
+				//TODO: because asking the blockchain if short_channel_id is valid is only optional
+				//in the blockchain API, we need to handle it smartly here, though it's unclear
+				//exactly how...
+				if checked_utxo {
+					// Either our UTXO provider is busted, there was a reorg, or the UTXO provider
+					// only sometimes returns results. In any case remove the previous entry. Note
+					// that the spec expects us to "blacklist" the node_ids involved, but we can't
+					// do that because
+					// a) we don't *require* a UTXO provider that always returns results.
+					// b) we don't track UTXOs of channels we know about and remove them if they
+					//    get reorg'd out.
+					// c) it's unclear how to do so without exposing ourselves to massive DoS risk.
+					Self::remove_channel_in_nodes(&mut self.nodes, &entry.get(), msg.contents.short_channel_id);
+					*entry.get_mut() = chan_info;
+				} else {
+					return Err(LightningError{err: "Already have knowledge of channel", action: ErrorAction::IgnoreError})
+				}
+			},
+			BtreeEntry::Vacant(entry) => {
+				entry.insert(chan_info);
+			}
+		};
+
+		macro_rules! add_channel_to_node {
+			( $node_id: expr ) => {
+				match self.nodes.entry($node_id) {
+					BtreeEntry::Occupied(node_entry) => {
+						node_entry.into_mut().channels.push(msg.contents.short_channel_id);
+					},
+					BtreeEntry::Vacant(node_entry) => {
+						node_entry.insert(NodeInfo {
+							channels: vec!(msg.contents.short_channel_id),
+							lowest_inbound_channel_fees: None,
+							announcement_info: None,
+						});
+					}
+				}
+			};
+		}
+
+		add_channel_to_node!(msg.contents.node_id_1);
+		add_channel_to_node!(msg.contents.node_id_2);
+
+		Ok(should_relay)
+	}
+
+	/// Close a channel if a corresponding HTLC fail was sent.
+	/// If permanent, removes a channel from the local storage.
+	/// May cause the removal of nodes too, if this was their last channel.
+	/// If not permanent, makes channels unavailable for routing.
+	pub fn close_channel_from_update(&mut self, short_channel_id: &u64, is_permanent: &bool) {
+		if *is_permanent {
+			if let Some(chan) = self.channels.remove(short_channel_id) {
+				Self::remove_channel_in_nodes(&mut self.nodes, &chan, *short_channel_id);
+			}
+		} else {
+			if let Some(chan) = self.channels.get_mut(&short_channel_id) {
+				if let Some(one_to_two) = chan.one_to_two.as_mut() {
+					one_to_two.enabled = false;
+				}
+				if let Some(two_to_one) = chan.two_to_one.as_mut() {
+					two_to_one.enabled = false;
+				}
+			}
+		}
+	}
+
+	fn fail_node(&mut self, _node_id: &PublicKey, is_permanent: &bool) {
+		if *is_permanent {
+			// TODO: Wholly remove the node
+		} else {
+			// TODO: downgrade the node
+		}
+	}
+
+	/// For an already known (from announcement) channel, update info regarding one of the directions of a channel.
+	/// Announcement signatures *are checked* here and should be checked by a caller beforehand.
```

Comment:
Why should they be checked beforehand if they're checked here? :p. I get the reasoning (the pubkey isnt available) but its super awkward to have it be different across different fns. Maybe we should just always do sig checking in NetworkGraph?

Reply:
But then in handle_channel_announcement we will be looking into the chain before checking sigs...
I guess it doesn't change much in this context.

Reply:
Hmm, ugh, good point. There's no winning here. Maybe just make the signature checking optional with both variants exposed for each and use the relevant one from the message handler?

Reply:
Technically, it's enough in this case to just make secp_ctx optional.

But maybe we want to force users to explicitly set `check_sig` variable, and then if secp_ctx is None, throw an exception?

Not sure which approach we should take within this project.

Reply:
Yea, I think provide-a-context-if-you-want-signatures-checked is pretty nice. just pass in an Option<&Secp256k1> and document it. No need to panic or force users to set a variable. good docs and a clear None/Some() distinction requires typing and thinking so I'm not too worried about it.